### DISCUSSION TITLE
Feature Suggestion - Ability to track modifications

### DISCUSSION THREAD

**Discussion by bfloon** [SMByC/ThRasE#5] [Ideas]
Hi 

I would like to request a feature for this plugin. I require the ability to track changes to a raster once I've used this plugin to modify it. It would be great to have the ability to save the polygons that have been drawn to modify the raster. or at least have some method of seeing the change that the raster has undergone versus the original. Perhaps some ability to see which individual cells / classes within the raster have changed would be good too. 

Thanks!

**Comment by XavierCLL** [comment]
Hi @bfloon 

Adding the option to track modifications done by the plugin is already a great idea! thanks for opening this suggestion. But I have some thoughts about it and I would like to know your opinion to discuss it:

- Mark the pixels changed is possible because the plugin knows that in every operation. But what do you think, what is the best way to mark those pixels without disturbing "visually" the image to the user? (here when the colours of the pixels are critical while the user is editing)
- Is saving the polygons drawn a good idea if the previous idea is implemented? I don't think so. And about that idea is partially good because there are different tools for editing (pixel, lines, freehand), and saving those (together) is more complicated

Thanks and regards!
Xavier

**Reply by bfloon** [reply]
hi @XavierCLL. Thank you very much for your reply. I think that the option to mark the pixels as changed would be a good idea- perhaps once the user is finished changing the raster they could have some method of exporting a 'change' raster that would show the pixels that had changed, and specifically what they had changed to. Or... perhaps the user could toggle between the changed layer as they are editing the raster, and then have the option to export this layer as required - this would prevent the visual disturbance. Not sure how viable this would be to implement, however. 

Thanks!

**Reply by XavierCLL** [reply]
Hi @bfloon, a raster layer to track the changes is not a good idea, but a vector layer is better, have a polygon (pixel) for every pixel changed and store in the table of the vector layer the old and the new value, and the user can visualize and save this layer. This option seems feasible to me

**Comment by xl7657** [comment]
Hi @bfloon and @XavierCLL, just noticed this interesting discussion here. I have tried several similar plugins to modify raster. There is a plugin called "Serval" (https://github.com/lutraconsulting/serval/blob/master/Serval/docs/user_manual.md) which allows users to draw polygons and then save the polygons. The main and the biggest issue of this plugin is user cannot change the pixel values from one to another.  It means everything in the polygons will be changed (can't target a specific pixel value). 

Pixel editor in ArcGIS pro (https://www.esri.com/arcgis-blog/products/arcgis-pro/imagery/pixel-editor-creating-useful-classified-raster-data/) seems do a pretty good job but same, users won't be able to track changes (once the pixel editor is closed, editing log will be removed). 

So what do you guys think if we draw polygons first, and then in the polygons we have field "current_class" and "new_class", and then change the raster based on that? So that we can target multiple pixels values at once in the same location. Not sure if this it is achievable, sounds very complicated to me :P 

Thank you guys!

**Reply by XavierCLL** [reply]
Hi @xl7657 

Serval plugin is for float images, and Thrase is only for categorical raster (with integer values, aka thematic) with several and specific features.

I am not sure if we are talking about the same thing here about "track modifications"...

What you mention at the end is one of the features of Thrase, using the recode pixel table to modify several pixel values at the same time in a specific region (polygon), is that or am I misunderstanding you?

Regards,
Xavier

**Comment by vieiramesquita** [comment]
Dear @bfloon and @XavierCLL, thank you for the amazing you guys developed. My feature suggestion it's relatively simple, would you guys provide supoort for multiple monitors/windows?

**Reply by XavierCLL** [reply]
Hi @vieiramesquita thanks! what exactly is your suggestion/discussion? Having multiple Thrase instances? (could you please move your new suggestion to another discussion thread with more details please) thanks

**Reply by vieiramesquita** [reply]
@XavierCLL, in addition to the possibility of splitting a window into columns and/or rows, is it possible to split it into multiple windows (for people who use multiple monitors)?

**Reply by XavierCLL** [reply]
@vieiramesquita now I understand, good idea, each window is a single view of Thrase, I will write down in the todo list.

Meanwhile, what you can do is split in two o three columns and expand the Thrase dialog to the multiple monitor and adjust it to the columns to the monitors (I know it is not the best, but it can work temporally)

**Comment by XavierCLL** [comment]
Hi @bfloon and other people interested in this feature.

Track modification was finally implemented! Check the new version https://github.com/SMByC/ThRasE/releases/tag/25.8

Let me know your comments
